cmd/tt/bucket.go

```diff
@@ -445,112 +536,232 @@ func addFile(key, name, filePath string) {
 		cmd.Fatal(err)
 	}
-	cmd.Message("Pushing %s to %s", aurora.White(name).Bold(), aurora.White(filePath).Bold())
 
-	bar := pbar.New(int(info.Size()))
-	bar.SetTemplate(pbar.Full)
-	bar.Set(pbar.Bytes, true)
-	bar.Set(pbar.SIBytesPrefix, true)
-	bar.Start()
+	bar := addBar(filePath, info.Size())
 	progress := make(chan int64)
-
 	go func() {
-		ctx, cancel := threadCtx(addFileTimeout)
-		defer cancel()
-		if _, _, err = buckets.PushPath(ctx, key, filePath, file, client.WithProgress(progress)); err != nil {
-			cmd.Fatal(err)
+		for up := range progress {
+			if err := bar.Set(int(up)); err != nil {
+				cmd.Fatal(err)
+			}
 		}
 	}()
 
-	for up := range progress {
-		bar.SetCurrent(up)
+	ctx, cancel := threadCtx(addFileTimeout)
+	defer cancel()
+	opts := []client.Option{client.WithProgress(progress)}
+	if !force {
+		opts = append(opts, client.WithFastForwardOnly(xroot))
 	}
-	bar.Finish()
+	added, root, err := buckets.PushPath(ctx, key, filePath, file, opts...)
+	if err != nil {
+		if strings.HasSuffix(err.Error(), bucks.ErrNonFastForward.Error()) {
+			cmd.Fatal(errors.New(nonFastForwardMsg), aurora.Cyan("tt bucket pull"))
+		} else {
+			cmd.Fatal(err)
+		}
+	} else {
+		finishBar(bar, filePath, added.Cid())
+	}
+	return root
 }
 
-var pullBucketPathCmd = &cobra.Command{
-	Use:   "pull [path] [destination]",
-	Short: "Pull a bucket path",
-	Long: `Pull files and directories from a bucket path. Existing paths will be overwritten. Non-existing paths will be created.
-
-Bucket structure is mirrored locally. For example, given the bucket:
-    foo/one.txt
-    foo/bar/two.txt
-    foo/bar/baz/three.txt
-
-These 'pull' commands result in the following local structures.
-
-'tt bucket pull foo mydir':
-    mydir/foo/one.txt
-    mydir/foo/bar/two.txt
-    mydir/foo/bar/baz/three.txt
-
-'tt bucket pull foo/bar mydir':
-    mydir/bar/two.txt
-    mydir/bar/baz/three.txt
-
-'tt bucket pull foo/bar/baz mydir':
-    mydir/baz/three.txt
-
-'tt bucket pull foo/bar/baz/three.txt mydir':
-    mydir/three.txt
+func rmFile(key string, xroot path.Resolved, filePath string, force bool) path.Resolved {
+	ctx, cancel := threadCtx(addFileTimeout)
+	defer cancel()
+	var opts []client.Option
+	if !force {
+		opts = append(opts, client.WithFastForwardOnly(xroot))
+	}
+	root, err := buckets.RemovePath(ctx, key, filePath, opts...)
+	if err != nil {
+		if strings.HasSuffix(err.Error(), bucks.ErrNonFastForward.Error()) {
+			cmd.Fatal(errors.New(nonFastForwardMsg), aurora.Cyan("tt bucket pull"))
+		} else if !strings.HasSuffix(err.Error(), "no link by that name") {
+			cmd.Fatal(err)
+		}
+	}
+	fmt.Println("- " + filePath)
+	return root
+}
 
-'tt bucket pull foo .':
-    foo/one.txt
-    foo/bar/two.txt
-    foo/bar/baz/three.txt
-`,
-	Args: cobra.ExactArgs(2),
+var bucketPullCmd = &cobra.Command{
+	Use:   "pull",
+	Short: "Pull bucket object changes",
+	Long:  `Pulls paths that have been added to and paths that have been removed or differ from the remote bucket root.`,
```

Comment:
^ same deal here... hopefully the code is somewhat readable.

Reply:
Just a general comment about this cli code, it may not be relevant because there is a lot to understand here and I definitely am not groking it all already...

Seems like there is a LOT of logic in these cli commands, as if the cli is actually implementing a type of buckets client. Just wondering if most of that logic can exist in an actual reusable client.

Reply:
Yeah, good point. I almost pulled everything into it's own sub package... would be nice to have tests at this level as well. Something for another day. I'll create an issue.